Pinta.Effects/Effects/ForwardErrorDiffusionDitheringEffect.cs

```diff
@@ -0,0 +1,372 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.ComponentModel;
+using System.Linq;
+using Cairo;
+using Pinta.Core;
+using Pinta.Gui.Widgets;
+
+namespace Pinta.Effects;
+
+public sealed class ForwardErrorDiffusionDitheringEffect : BaseEffect
+{
+	public override string Name => Translations.GetString ("Dithering");
+	public override bool IsConfigurable => true;
+	// TODO: Icon
+	public override string EffectMenuCategory => Translations.GetString ("Color");
+	public ForwardErrorDiffusionDitheringData Data => (ForwardErrorDiffusionDitheringData) EffectData!; // NRT - Set in constructor
+
+	public override bool IsTileable => false;
+
+	public ForwardErrorDiffusionDitheringEffect ()
+	{
+		EffectData = new ForwardErrorDiffusionDitheringData ();
+	}
+
+	public override void LaunchConfiguration ()
+	{
+		EffectHelper.LaunchSimpleEffectDialog (this);
+	}
+
+	private sealed record DitheringSettings (
+		ErrorDiffusionMatrix diffusionMatrix,
+		ImmutableArray<ColorBgra> palette,
+		int sourceWidth,
+		int sourceHeight);
+
+	private DitheringSettings CreateSettings (ImageSurface src)
+		=> new (
+			diffusionMatrix: GetPredefinedDiffusionMatrix (Data.DiffusionMatrix),
+			palette: GetPredefinedPalette (Data.PaletteChoice),
+			sourceWidth: src.Width,
+			sourceHeight: src.Height
+		);
+
+	protected override void Render (ImageSurface src, ImageSurface dest, RectangleI roi)
+	{
+		DitheringSettings settings = CreateSettings (src);
+
+		Span<ColorBgra> dst_data = dest.GetPixelData ();
+
+		for (int y = roi.Top; y <= roi.Bottom; y++) {
+
+			for (int x = roi.Left; x <= roi.Right; x++) {
+
+				int currentIndex = y * settings.sourceWidth + x;
+				ColorBgra originalPixel = dst_data[currentIndex];
```

Comment:
I think you need to be reading from the `src` image surface here - the destination isn't necessarily a copy of the image already. I'm pretty sure this is causing the test failure since the unit test creates a blank surface to contain the result